### PR TITLE
Third time's the charm - Photography update: 7x7 cameras, photo logging with full metadata, persistent albums and wall frames!

### PR DESCRIPTION
**Pull request by silicons**
Why: Because cameras need more love, and I've sped up getFlatIcon so why not add more lag to it?? (just kidding, please do not take the latter part seriously). Also, persistant photos would be a fun thing!

:cl:
rscadd: Cameras now shoot from 1x1 to 7x7, fully customizable. Alt click on them to change the size of your photos.
experimental: All photos, assuming the server host turns this feature on, will be logged to disk in round logs, with their data and path stored in a json. This allows for things like Statbus, and persistence features among other things to easily grab the data and load the photo.
rscadd: Mappers are now able to add in photo albums and wall frames with persistence! This, obviously, requires photo logging to be turned on. If this is enabled and used, these albums and frames will save the ID of the photo(s) inside them and load it the next time they're loaded in! Like secret satchels, but for photos!
/:cl:

### CODE REVIEW

**Comment by Mickyan** [comment]
The dream is alive once more

I've been wondering, if you set the persistence id to be the user's ckey (spawned through the library console for example) you could theoretically have personal photo albums right? Or is having a potentially infinite number of persistent albums an issue for filesize?

**Comment by kevinz000** [comment]
@Mickyan All the persistence does is save the ID of the album/frame, and the id(s) of the things inside. The disk storage would be used by storing the actual photos in the log directories, that the ids reference.
Basically, you can have basically as much persistence as you want, as the storage is already being used.
Persistence, however, will require all photo logging to be turned on. Can be tweaked in the future/whenever maintainers/ @MrStonedOne tells me to stop wasting disk space =p

**Comment by Cyberboss** [comment]
>storing binary data in the log directory

Let me just stop you right there

**Comment by epochayur** [comment]
should there be some sort of message in the description like "Alt-click to change photo size" or something like that? i feel like it's not terribly obvious

**Comment by Cyberboss** [comment]
fix conflict or i stale

**Comment by PKPenguin321** [comment]
Here's to hoping this gets through this time

**Comment by ShizCalev** [comment]
compile

File: code/modules/photography/camera/camera_image_capturing.dm [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		var/icon/img = getFlatIcon(A)
+		res.Blend(img, blendMode2iconMode(A.blend_mode), xo, yo)
+		CHECK_TICK
+
+	if(istype(custom_sound))				//This is where the camera actually finishes its exposure.
+		playsound(loc, custom_sound, 75, 1, -3)
+	else
+		playsound(loc, pick('sound/items/polaroid1.ogg', 'sound/items/polaroid2.ogg'), 75, 1, -3)
+
+	//if(wipe_atoms)
```
Comment (ShizCalev) [inline code/modules/photography/camera/camera_image_capturing.dm]: ded code

**Comment by kevinz000** [comment]
working on it

**Comment by kevinz000** [comment]
AIGHT, DONE.

File: code/modules/photography/camera/camera.dm [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+	var/blueprints = FALSE
+	var/clone_area = SSmapping.RequestBlockReservation(size_x * 2 + 1, size_y * 2 + 1)
+	for(var/turf/T in block(locate(target_turf.x - size_x, target_turf.y - size_y, target_turf.z), locate(target_turf.x + size_x, target_turf.y + size_y, target_turf.z)))
+		if((ai_user && GLOB.cameranet.checkTurfVis(T)) || T in seen)
+			turfs += T
+			for(var/mob/M in T)
+				mobs += M
+			if(locate(/obj/item/areaeditor/blueprints) in T)
+				blueprints = TRUE
+			CHECK_TICK
```
Comment (Cyberboss) [inline code/modules/photography/camera/camera.dm]: The whole point of the reservation thing was so this CHECK_TICK wouldn't be necessary

**Comment by Mickyan** [comment]
Getting the pop up to customize the text every time you take a photo is really annoying, is there a reason why it needs to pop up automatically instead of using a pen on it

File: code/modules/photography/camera/camera_image_capturing.dm [hunk 3]
```diff
@@ -0,0 +1,88 @@
+/obj/effect/appearance_clone
+
+/obj/effect/appearance_clone/New(loc, atom/A)			//Intentionally not Initialize().
```
Comment (Cyberboss) [inline code/modules/photography/camera/camera_image_capturing.dm]: What reason could you possibly have for this?

Comment (kevinz000) [reply code/modules/photography/camera/camera_image_capturing.dm]: Didn't know if initialize was "immediate" enough to let the camera call getFlatIcon on it immediately.

**Comment by kevinz000** [comment]
@Mickyan it's saved to disk after you do the popup stuff that's why

**Comment by TheDreamweaver** [comment]
Would it be possible to get photos to capture lighting

File: code/modules/photography/camera/camera_image_capturing.dm (hunk 3 above)
Comment (optimumtact) [reply code/modules/photography/camera/camera_image_capturing.dm]: maybe you should include that in the comment

File: code/modules/photography/_pictures.dm [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	var/caption
+	var/icon/picture_image
+	var/icon/picture_icon
+	var/psize_x = 0
+	var/psize_y = 0
+	var/has_blueprints = FALSE
+	var/logpath						//If the picture has been logged this is the path.
+	var/id							//this var is NOT protected because the worst you can do with this that you couldn't do otherwise is overwrite photos, and photos aren't going to be used as attack logs/investigations anytime soon.
+
+/datum/picture/New(name = "picture", desc = "This is a picture. A bugged one. Report it to coderbus!", image, icon, size_x = 96, size_y = 96, bp = FALSE, caption_, autogenerate_icon = FALSE)
```
Comment (AnturK) [inline code/modules/photography/_pictures.dm]: Why are these different than default values.

File: code/modules/photography/camera/camera.dm [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+	var/picture_size_x_max = 3
+	var/picture_size_y_max = 3
+
+/obj/item/camera/CheckParts(list/parts_list)
+	..()
+	var/obj/item/camera/C = locate(/obj/item/camera) in contents
+	if(C)
+		pictures_max = C.pictures_max
+		pictures_left = C.pictures_left
+		visible_message("[C] has been imbued with godlike power!")
```
Comment (AnturK) [inline code/modules/photography/camera/camera.dm]: ??? Even if this is about some blessing craft, this is assuming everything done on cameras does this.

File: code/modules/photography/camera/camera.dm [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+	var/on = TRUE
+	var/cooldown = 64
+	var/blending = FALSE		//lets not take pictures while the previous is still processing!
+	var/see_ghosts = CAMERA_NO_GHOSTS //for the spoop of it
+	var/obj/item/disk/holodisk/disk
+	var/sound/custom_sound
+	var/silent = FALSE
+	var/picture_size_x = 1
+	var/picture_size_y = 1
+	var/picture_size_x_min = 0
```
Comment (AnturK) [inline code/modules/photography/camera/camera.dm]: Why is minimum zero, and if it means 1 why show 0 to users.

File: code/modules/photography/camera/camera.dm [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+			to_chat(user, "<span class='warning'>There's already a disk inside [src].</span>")
+		return TRUE //no afterattack
+	..()
+
+/obj/item/camera/examine(mob/user)
+	..()
+	to_chat(user, "It has [pictures_left] photos left.")
+
+/obj/item/camera/proc/can_target(atom/target, mob/user, prox_flag)
+	if(!on || blending || !pictures_left || (!isturf(target) && !isturf(target.loc)) || !((isAI(user) && GLOB.cameranet.checkTurfVis(get_turf(target))) || ((user.client && (get_turf(target) in get_hear(user.client.view, user)) || (get_turf(target) in get_hear(world.view, user))))))
```
Comment (AnturK) [inline code/modules/photography/camera/camera.dm]: Split this please, it's not performance issue and it looks horrible to read.

File: code/modules/photography/photos/album.dm [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+	var/list/current_ids = get_picture_id_list()
+	for(var/i in ids)
+		if(i in current_ids)
+			continue
+		var/obj/item/photo/P = load_photo_from_disk(i)
+		if(istype(P))
+			if(!SEND_SIGNAL(src, COMSIG_TRY_STORAGE_INSERT, P, null, TRUE, TRUE))
+				qdel(P)
+
+#define ALBUM_DEFINE(id) /obj/item/storage/photo_album/##id/persistence_id = #id
```
Comment (AnturK) [inline code/modules/photography/photos/album.dm]: No.

File: code/modules/photography/camera/camera.dm [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+/obj/item/camera/proc/camera_get_mobdesc(mob/M)
+	var/list/mob_details = list()
+	if(M.invisibility)
+		if(see_ghosts && isobserver(M))
+			mob_details += "You can also see a g-g-g-g-ghooooost! "
+		else
+			return mob_details
+	var/list/holding = list()
+	if(isliving(M))
+		var/mob/living/L = M
```
Comment (AnturK) [inline code/modules/photography/camera/camera.dm]: You could probably abstract these into some /get_photo_description for both ghosts and other mobs.

**Review by AnturK — CHANGES_REQUESTED** [review]

**Comment by PKPenguin321** [comment]
:tada:

**Pull request merged**